python/src/cloudstorage/storage_api.py

```diff
@@ -169,6 +169,42 @@ def get_bucket_async(self, path, **kwds):
     """GET a bucket."""
     return self.do_request_async(self.api_url + path, 'GET', **kwds)
 
+  # pylint: disable=too-many-locals
+  def compose_object(self, file_list, destination_file, content_type):
+    """COMPOSE multiple objects together.
+
+    Taking a list of files it calls the put object with the compose flag
+      this call merges all the files into the destination file
+    Currently stubbed because the dev server cloudstorage_stub.py
+      does not handle compose requests.
+    TODO: When the dev server gets patch please remove the stub
+
+    Args:
+      file_list: file_list: list of dicts with the file name.
+      destination_file: Path to the destination file.
+      content_type: Content type for the destination file.
+      retry_params: An api_utils.RetryParams for this call to GCS. If None,
+      the default one is used.
+    _account_id: Internal-use only.
+    """
+    xml = ''
+    for meta_data in file_list:
+      xml_meta = ""
+      for meta in meta_data:
+        xml_meta += "<%s>%s</%s>" % (meta, meta_data[meta], meta)
+      xml += '<Component>%s</Component>' % xml_meta
+    xml = '<ComposeRequest>%s</ComposeRequest>' % xml
```

Comment:
On Mon, Mar 30, 2015 at 1:44 PM, rbruyere <notifications@github.com> wrote:

> In python/src/cloudstorage/storage_api.py
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6#discussion_r27429987>
> :
>
> > +
> > +    Args:
> > +      file_list: file_list: list of dicts with the file name.
> > +      destination_file: Path to the destination file.
> > +      content_type: Content type for the destination file.
> > +      retry_params: An api_utils.RetryParams for this call to GCS. If None,
> > +      the default one is used.
> > +    _account_id: Internal-use only.
> > +    """
> > +    xml = ''
> > +    for meta_data in file_list:
> > +      xml_meta = ""
> > +      for meta in meta_data:
> > +        xml_meta += "<%s>%s</%s>" % (meta, meta_data[meta], meta)
> > +      xml += '<Component>%s</Component>' % xml_meta
> > +    xml = '<ComposeRequest>%s</ComposeRequest>' % xml
>
> How is it not efficient. Any recommendations?
>
In general a + b + c would be better than (a + "%s" + c) % b  (think from
allocated bytes and byte iteration perspective).
So, in your case plain initialization xml_meta with "<Component>" and
adding "</<Component>" after iteration and doing
the same for xml would be way more efficient (and without doing anything
special).
If you really want to be efficient you should consider (cStringIO or join)
- see:  https://waymoot.org/home/python_string/

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6/files#r27429987>
> .
>